driver-core/src/main/java/com/datastax/driver/core/CodecRegistry.java

```diff
@@ -219,7 +201,65 @@ public int hashCode() {
         }
     }
 
-    private final boolean immutable;
+    /**
+     * A complexity-based weigher.
+     * Weights are computed according to the CQL type:
+     * <ol>
+     * <li>Primtive types weigh 0;
+     * <li>Collections weigh the total weight of their inner types + the weight of their level of deepness;
+     * <li>UDTs and tuples weigh the total weight of their inner types + the weight of their level of deepness, but cannot weigh less than 1;
+     * <li>Custom (non-CQL) types weigh 1.
+     * </ol>
+     * A consequence of this algorithm is that primitive types and all "shallow" collections thereof
+     * are never evicted.
+     */
+    private static class TypeCodecWeigher implements Weigher<CacheKey, TypeCodec<?>> {
```

Comment:
That sounds great, that looks like it'd work perfectly!  I think our only concern is if in theory somewhere registers a crazy amount of codecs themselves, but looks like that would take 582 registered codecs, so we probably don't need to expect this right?   Another concern is if we add enough types in future releases, we'll just have to be mindful of that capacity (I don't expect this to ever be a problem).